jax/experimental/mosaic/gpu/utils.py

```diff
@@ -1115,6 +1124,80 @@ def memref_ptr(memref_arg, memory_space=None):
   )
 
 
+def extern_elementwise_func_apply(func, return_type, arg_list,
+                                  static_arg_idxs=(), pack_elts=False,
+                                  elt_byte_count_idx=None):
+  tidx = thread_idx()
+  wgidx = warpgroup_idx()
+  index = ir.IndexType.get()
+  i64 = ir.IntegerType.get_signless(64)
+  i32 = ir.IntegerType.get_signless(32)
+  ptr_ty = ir.Type.parse("!llvm.ptr")
+  num_elts = None
+  elem_bytewidth = None
+  num_args = len(arg_list)
+  tensor_arg_idxs = set(range(num_args)) - set(static_arg_idxs)
+  wg_thread_offset = arith.muli(wgidx, c(WARPGROUP_SIZE, i32))
+  wg_tidx = arith.subi(tidx, wg_thread_offset)
+  num_cuda_threads = WARPGROUP_SIZE
+
+  arg_list_ptr_int = arg_list.copy()
+  for idx in tensor_arg_idxs:
+    arg = arg_list_ptr_int[idx]
+    if num_elts is None:
+      ref_ty = ir.MemRefType(arg.type)
+      _, slice_shape, _ = parse_indices((), ref_ty.shape)
+      num_elts = np.prod(slice_shape)
+      elem_bytewidth = bytewidth(ref_ty.element_type)
+
+    arg_list_ptr_int[idx] = llvm.ptrtoint(i64, memref_ptr(arg))
+
+  def run_func(wave_idx, num_elt_bytes):
+    args = arg_list_ptr_int.copy()
+    # Update element arg ptrs
+    for idx in tensor_arg_idxs:
+      offset_elems = arith.addi(arith.index_cast(index, wg_tidx),
+                                arith.muli(wave_idx, c(num_cuda_threads, index)))
+      offset_bytes = llvm.mul(
+        arith.index_cast(i64, offset_elems),
+        c(num_elt_bytes, i64),
+        overflow_flags=llvm.IntegerOverflowFlags.none,
+      )
+      args[idx] = llvm.inttoptr(
+        ptr_ty,
+        llvm.add(
```

Comment:
nit: you can use `arith.add` to avoid having to specify the overflow flags all the time